### PR TITLE
feat: Filter workflows by "Finished before" and "Created Since" via API. Fixes #13151

### LINKED ISSUE #13151: UI: "Finished before" field only affects visible Workflows
### Pre-requisites

- [X] I have double-checked my configuration
- [X] I have tested with the `:latest` image tag (i.e. `quay.io/argoproj/workflow-controller:latest`) and can confirm the issue still exists on `:latest`. If not, I have explained why, **in detail**, in my description below.
- [X] I have searched existing issues and could not find a match for this bug
- [ ] I'd like to contribute the fix myself (see [contributing guide](https://github.com/argoproj/argo-workflows/blob/main/docs/CONTRIBUTING.md))

### What happened/what did you expect to happen?

In 3.5.7, the "Finished before" field in the UI doesn't work properly. It only filters and displays results from the first page of workflows. This means if you enter a date that is before the dates of the workflows on first page, nothing shows up.

image
image
image

### Version

3.5.7

### Paste a small workflow that reproduces the issue. We must be able to run the workflow; don't enter a workflows that uses private images.

```YAML
N/A
```

### Logs from the workflow controller

```text
N/A
```

### Logs from in your workflow's wait container

```text
N/A
```

### PR DESCRIPTION
**Pull request by Adrien-D**
Fixes #13151

### Motivation

- "Finished before" and "Created Since" are filtering workflows client-side so not all workflows are shown

### Modifications

- I implemented `createdAfter` and `finishedBefore` on server side similarly to https://github.com/argoproj/argo-workflows/issues/12161.
- Remark on implementation: `createdAfter` is based on `metadata.creationTimestamp` which is stored as json so I needed to extract it in the request [here](https://github.com/argoproj/argo-workflows/compare/main...Adrien-D:filter-finished-before?expand=1#diff-f0e0e995eed316a255ab70a36d5c06588bd89c73a2a0dabac9bbbfa34b6df5aeR59). Not sure about any perf impact on this, but I didn't find any other simple way of doing it.
- I didn't update the CLI as mentioned [here](https://github.com/argoproj/argo-workflows/issues/13151#issuecomment-2153110359) by @agilgur5 as I never worked with the CLI and don't really know what to do. Can someone give me an insight on that or maybe we could split this in another issue to implement in the CLI later ?

### Verification

- Tests for `ListWokflows` function
- Local testing on existing worflows

| Initial workflows  | screenshot  |
|---|---|
| Filtering created since | screenshot |
| Filtering finished before | screenshot |
| Filtering finished before | screenshot |

### CODE REVIEW

**Comment by agilgur5** [comment]
> * I didn't update the CLI as mentioned [here](https://github.com/argoproj/argo-workflows/issues/13151#issuecomment-2153110359) by @agilgur5 as I never worked with the CLI and don't really know what to do. Can someone give me an insight on that [...]?

The CLI is one of the most straightforward parts of the codebase (simpler than the UI and API, IMO); each command has a file (e.g. `cmd/argo/commands/list.go`) with around ~100-200 LoC each (depending on the number of flags).
You can see the existing client-side date filters in the `list` command [starting on this line](https://github.com/argoproj/argo-workflows/blob/97b94f0f67714beacd7b6f2ef086d13f09cd158f/cmd/argo/commands/list.go#L184)

**Comment by agilgur5** [comment]
> Verification

you'll want to test when both are set as well since there are some edge cases with that, although now that these are in the API, that can be covered with back-end tests

and ensure that with and without archived workflows works correctly

**Comment by tooptoop4** [comment]
@agilgur5  join the band again 🎸

**Comment by agilgur5** [comment]
My original rationale has not changed, so that won't happen, but if I'm tagged somewhere (as I was here) and it's related to something I wrote, I do try to be helpful as a general principle.
I didn't do an in-depth review, but as I was looking through anyway, I reviewed a small bit

File: server/workflowarchive/archived_workflow_server.go [hunk 1]
```diff
@@ -45,7 +45,7 @@ func (w *archivedWorkflowServer) ListArchivedWorkflows(ctx context.Context, req
 		listOptions = *req.ListOptions
 	}
 
-	options, err := sutils.BuildListOptions(listOptions, req.Namespace, req.NamePrefix, "")
+	options, err := sutils.BuildListOptions(listOptions, req.Namespace, req.NamePrefix, "", "", "")
```
[the remark this answers is not in the document]
Comment (Adrien-D) [reply server/workflowarchive/archived_workflow_server.go:48]: Indeed I missed that part for archived workflows, I'll change that and add tests.

For `minStartedAt` and `maxStartedAt` I didn't used those because I wasn't sure if they were still used somehow, and because they are not based on the same attribute so the query will be different. I can rename in the API if you tell me it's safe.

[the remark this answers is not in the document]
Comment (Adrien-D) [reply server/workflowarchive/archived_workflow_server.go:48]: I tried to update the code like this, but it seems that `createdAfter` and `finishedBefore` have no effect, not sure they are passed down to the API call. I spent few time trying to understand what's going on, do you have any idea ? I must miss something 

```golang
var workflows wfv1.Workflows
createdAfter := ""
finishedBefore := ""
if flags.createdSince != "" {
	createdTime, err := argotime.ParseSince(flags.createdSince)
	if err != nil {
		return nil, err
	}
	createdAfter = createdTime.Format(time.RFC3339)
}
if flags.finishedBefore != "" {
	finishedTime, err := argotime.ParseSince(flags.createdSince)
	if err != nil {
		return nil, err
	}
	finishedBefore = finishedTime.Format(time.RFC3339)
}

for {
	log.WithField("listOpts", listOpts).Debug()
	wfList, err := serviceClient.ListWorkflows(ctx, &workflowpkg.WorkflowListRequest{
		Namespace:      flags.namespace,
		ListOptions:    listOpts,
		Fields:         flags.displayFields(),
		CreatedAfter:   createdAfter,
		FinishedBefore: finishedBefore,
	})
	if err != nil {
		return nil, err
	}
	workflows = append(workflows, wfList.Items...)
	if wfList.Continue == "" {
		break
	}
	listOpts.Continue = wfList.Continue
}
```

[the remark this answers is not in the document]
Comment (tooptoop4) [reply server/workflowarchive/archived_workflow_server.go:48]: 🚢 📦

**Review by terrytangyuan — APPROVED** [review]

**Review by rohankmr414 — APPROVED** [review]
LGTM

**Comment by meydominic** [comment]
Hey,
I need to ask something related to this pr. I saw that the workflow.pb.go was adapted for the filters. But there were no adaptions on the workflow-archive.pb.go.

All our workflows will be put to archive after they ran. A client can have thousands of ran workflows after a few weeks. (For example, one of our clients has run about 27k workflows in the last four weeks - and this is only from a testing phase). So loading all workflows and filter them afterwards is no option 😆 

So if I am right, this feature will not add the finishedBefore and createdSince filters to archived workflows? If yes - how is the correct way to filter and fetch them without loading them all first.

Thanks and best regards,
Dominic
